src/audio/smart_amp_aw/aw_smart_amp_wapper.c

```diff
@@ -0,0 +1,758 @@
+
+	/* buffer : sof -> frame_iv memory allocation */
+	size = SMART_AMP_FB_BUF_DB_SZ * sizeof(int32_t);
+	sktune->frame_iv.data_ptr = rballoc(0, SOF_MEM_CAPS_RAM, size);
+	if (!sktune->frame_iv.data_ptr) {
+		comp_err(dev, "[Awinic] frame_iv alloc failed");
+		sktune->frame_iv.actual_data_len = 0;
+		sktune->frame_iv.max_data_len = 0;
+		goto err;
+	}
+	sktune->frame_iv.actual_data_len = 0;
+	sktune->frame_iv.max_data_len = size;
+
+	sum_mem_size += size;
+
+
+	/* memory allocation of SKTune handle */
+	algo_size = sktune_api_get_size();
+	if(algo_size == 0) {
+		comp_err(dev, "[Awinic] get memory failed, algo_size =  %d", algo_size);
+		goto err;
+	}
+
+	sktune->handle = rballoc(0, SOF_MEM_CAPS_RAM, algo_size);
+	if (!sktune->handle) {
+		comp_err(dev, "[Awinic] SKTune handle alloc failed");
+		goto err;
+	}
+
+	sum_mem_size += (int)algo_size;
+
+	comp_dbg(dev, "[Awinic] module:%p (%d bytes used)", sktune, sum_mem_size);
+
+	return sktune;
+
+err:
+	smart_amp_sktune_free(sktune);
+	rfree(sktune);
+	return NULL;
+
+}
+
+
+int smart_amp_init(sktune_t *      sktune, struct comp_dev *dev)
+{
+	int ret;
+
+	ret = sktune_api_init(sktune->handle);
+	if (ret < 0) {
+		comp_err(dev, "[Awinic] SKTune init failed");
+		return ret;
+	}
+
+	sktune_api_set_media_info(sktune->handle, (void *)&sktune->media_info);
+
+	sktune->sub_msg_info[AW_MSG_ID_0].status = AW_MESG_NONE;
+	sktune->sub_msg_info[AW_MSG_ID_1].status = AW_MESG_NONE;
+	sktune->sub_msg_info[AW_MSG_ID_0].opcode_id = 0;
+	sktune->sub_msg_info[AW_MSG_ID_1].opcode_id = 0;
+
+	return 0;
+}
+
+int smart_amp_deinit(sktune_t *       sktune, struct comp_dev *dev)
+{
+	int ret;
+
+	ret = sktune_api_end(sktune->handle);
+	if (ret < 0) {
+		comp_err(dev, "[Awinic] Awinic End  failed");
+	}
+
+	smart_amp_sktune_free(sktune);
+
+	return 0;
+}
+
+
+int smart_amp_check_audio_fmt(int sample_rate, int ch_num)
+{
+
+	if (sample_rate != 4800) {
+		return -1;
+	}
+
+	if (ch_num != 2) {
+		return -1;
+	}
+
+	return 0;
+}
+
+int smart_amp_flush(sktune_t *sktune, struct comp_dev *dev)
+{
+	memset(sktune->frame_in.data_ptr, 0,
+	       SMART_AMP_FF_BUF_DB_SZ * sizeof(int32_t));
+	memset(sktune->frame_out.data_ptr, 0,
+	       SMART_AMP_FF_BUF_DB_SZ * sizeof(int32_t));
+	memset(sktune->frame_iv.data_ptr, 0,
+	       SMART_AMP_FB_BUF_DB_SZ * sizeof(int32_t));
+
+
+	comp_dbg(dev, "[Awinic] Reset (handle:%p)", sktune);
+
+	return 0;
+}
+
+
+static int smart_amp_get_buffer(char *buf, uint32_t frames,
+				const struct audio_stream __sparse_cache *stream, uint32_t num_ch)
+{
+	int idx;
+	uint32_t sample_num = frames * num_ch;
+	int byte_num = 0;
+	int16_t *input16, *output16;
+	int32_t *input32, *output32;
+
+	input16 = audio_stream_get_rptr(stream);
+	input32 = audio_stream_get_rptr(stream);
+	output16 = (int16_t *)buf;
+	output32 = (int32_t *)buf;
+
+	switch (audio_stream_get_frm_fmt(stream)) {
+	case SOF_IPC_FRAME_S16_LE:
+		for (idx = 0 ; idx < sample_num ; idx++) {
+				input16 = audio_stream_read_frag_s16(stream, idx);
```

Comment:
@singalsu can you comment on the proper way to read the buffer? this looks like the legacy method

Reply:
Is this copy from circular to linear buffer? There's a function available for that.